tests/rewards/test_reward_nets.py

```diff
@@ -531,30 +520,89 @@ def test_add_std_reward_wrapper(
     numpy_transitions: NumpyTransitions,
 ):
-    basic_reward_net = reward_nets.BasicRewardNet(
-        env_2d.observation_space,
-        env_2d.action_space,
-    )
-    basic_reward_net.predict_processed = mock.Mock(return_value=np.zeros((10,)))
-    wrapped_reward_net = make_wrapper(
-        basic_reward_net,
+    def potential(x: th.Tensor):
+        return th.full((x.shape[0],), 10, device=x.device)
+
+    shaped = reward_nets.ShapedRewardNet(zero_reward_net, potential, 0.9)
+    # We expect the shaped reward to be -1 since,
+    # r'(s,a,s') = r(s,a,s') + \gamma \theta(s') - \theta(s) = (0) + (0.9)(10) - 10 = -1
+    shaped_rew = th.full((10,), -1, dtype=th.float32)
+    forward_args = shaped.preprocess(*numpy_transitions)
+    assert th.allclose(shaped(*forward_args), shaped_rew)
+    assert th.allclose(shaped.predict_th(*numpy_transitions), shaped_rew)
+    assert np.allclose(shaped.predict(*numpy_transitions), shaped_rew.numpy())
+    assert np.allclose(shaped.predict_processed(*numpy_transitions), shaped_rew.numpy())
+
+
+@pytest.mark.parametrize("make_forward_wrapper", MAKE_FORWARD_WRAPPERS)
+def test_forward_wrapper_cannot_be_applied_predict_processed_wrapper(
+    predict_processed_wrapper: reward_nets.PredictProcessedWrapper,
+    make_forward_wrapper: MakeForwardWrapper,
+):
+    with pytest.raises(
+        ValueError,
+        match=r"ForwardWrapper cannot be applied on top of PredictProcessedWrapper!",
+    ):
+        make_forward_wrapper(predict_processed_wrapper)
+
+
+@pytest.mark.parametrize(
+    "make_predict_processed_wrapper",
+    MAKE_PREDICT_PROCESSED_WRAPPERS,
+)
+def test_predict_processed_wrappers_pass_on_kwargs(
+    make_predict_processed_wrapper: MakePredictProcessedWrapper,
+    zero_reward_net: MockRewardNet,
+    numpy_transitions: NumpyTransitions,
+):
+    zero_reward_net.predict_processed = mock.Mock(return_value=np.zeros((10,)))
+    wrapped_reward_net = make_predict_processed_wrapper(
+        zero_reward_net,
     )
     wrapped_reward_net.predict_processed(
         *numpy_transitions,
         foobar=42,
     )
-    basic_reward_net.predict_processed.assert_called_once_with(
+    zero_reward_net.predict_processed.assert_called_once_with(
         *numpy_transitions,
         foobar=42,
     )
 
 
+def test_predict_processed_wrappers_to_pass_on_method_calls_to_base(
+    numpy_transitions: NumpyTransitions,
+    torch_transitions: TorchTransitions,
+    predict_processed_wrapper: reward_nets.PredictProcessedWrapper,
+):
+    base = mock.create_autospec(MockRewardNet)
+
+    base.device = th.device("cpu")
+
+    base.dtype = th.float32
+
+    predict_processed_wrapper._base = base
+    # Check method calls
+    for attr, call_with, return_value in [
+        ("forward", torch_transitions, th.zeros(10)),
+        ("predict_th", numpy_transitions, np.zeros(10)),
+        ("predict", numpy_transitions, np.zeros(10)),
+        ("predict_processed", numpy_transitions, np.zeros(10)),
```

Comment:
I'm confused why this line passes. `PredictProcessedWrapper` does not have a concrete implementation of `predict_processed`, so shouldn't line 597 fail with this?

Reply:
We're never actually creating instances of `PredictProcessedWrapper` directly since it's an abstract class. The fixture `predict_processed_wrapper` is parametrized, so it actually tests all subclasses of `PredictProcessedWrapper` listed in `MAKE_PREDICT_PROCESSED_WRAPPER`.

Thus, this test just checks that the subclasses of `PredictProcessedWrapper` maintain certain properties. For example calling down to the base classes `predict_processed` method when calling its own `predict_processed` method. Whether it is unreasonable to enforce these for all future subclasses is another question.